tree/treeplayer/inc/ROOT/TVec.hxx

```diff
@@ -0,0 +1,574 @@
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "==", [](const T &t, const V &v) -> Boolean_t { return t == v; });
+}
+
+template <typename T, typename V>
+TVec<Boolean_t> operator!=(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "!=", [](const T &t, const V &v) -> Boolean_t { return t != v; });
+}
+template <typename T, typename V>
+TVec<Boolean_t> operator<=(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "<=", [](const T &t, const V &v) -> Boolean_t { return t <= v; });
+}
+
+template <typename T, typename V>
+TVec<Boolean_t> operator<(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "<", [](const T &t, const V &v) -> Boolean_t { return t < v; });
+}
+
+template <typename T, typename V>
+TVec<Boolean_t> operator&&(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "&&", [](const T &t, const V &v) -> Boolean_t { return t && v; });
+}
+
+template <typename T, typename V>
+TVec<Boolean_t> operator||(const TVec<T> &v0, const TVec<V> &v1)
+{
+   return ROOT::Internal::VecOps::Operate(v0, v1, "||", [](const T &t, const V &v) -> Boolean_t { return t || v; });
+}
+
+///@}
+
+/** @name Math Functions
+ *  Math functions on TVecs
+*/
+///@{
+#define MATH_FUNC(FUNCNAME)                                                                   \
+   template <typename T>                                                                      \
+   auto FUNCNAME(const TVec<T> &v)->TVec<decltype(std::FUNCNAME(v[0]))>                       \
+   {                                                                                          \
+      return ROOT::Internal::VecOps::Operate(v, [](const T &t) { return std::FUNCNAME(t); }); \
+   }
+
+MATH_FUNC(sqrt)
+MATH_FUNC(log)
+MATH_FUNC(sin)
+MATH_FUNC(cos)
+MATH_FUNC(asin)
+MATH_FUNC(acos)
+MATH_FUNC(tan)
+MATH_FUNC(atan)
+MATH_FUNC(sinh)
+MATH_FUNC(cosh)
+MATH_FUNC(asinh)
+MATH_FUNC(acosh)
+MATH_FUNC(tanh)
+MATH_FUNC(atanh)
+MATH_FUNC(abs)
+
+///@}
+
+/// Inner product
+template <typename T, typename V>
+auto Dot(const TVec<T> &v0, const TVec<V> &v1) -> decltype(v0[0] * v1[0])
+{
+   ROOT::Internal::VecOps::CheckSizes(v0.size(), v1.size(), "Dot");
+   return std::inner_product(v0.begin(), v0.end(), v1.begin(), decltype(v0[0] * v1[0])(0));
+}
+
+/// Sum elements
+template <typename T>
+T Sum(const TVec<T> &v)
+{
+   return std::accumulate(v.begin(), v.end(), T(0));
+}
+
+/// Create new collection applying a callable to the elements of the input collection
+template <typename T, typename F>
+auto Map(const TVec<T> &v, F &&f) -> TVec<decltype(f(v[0]))>
+{
+   return ROOT::Internal::VecOps::Operate(v, std::forward<F>(f));
+}
+
+/// Create a new collection with the elements passing the filter expressed by the predicate
+template <typename T, typename F>
+TVec<T> Filter(const TVec<T> &v, F &&f)
```

Comment:
Should we also have an operation that uses a mask?

Reply:
Isn't this &&? For example: mu_pt && (abs(mu_eta) < 2)

Reply:
Isn't that *resulting* in a mask? What if I have such a mask and I'd like a vector that is possibly smaller, selecting only those elements from the original vector that pass the mask? I.e. in `Filter()`s terms:
```
auto mask = Map(v, f);
assert(mask.size() == f.size());
auto reduced = ThisOpMightBeMissing_CouldBeCalledReduce(v, mask);
```
instead of
```
auto reduced = Filter(v, f);
```

Reply:
I see (with 48h delay, sorry).
This is what is available with the overloaded operator[], like in numpy:
```
auto reduced = v[mask];
```
or with a Physics hat on top:
```
auto goodPts = pts[pts > 15 && abs(etas) < 1.2];
```